p2p/types/src/network_address.rs

```diff
@@ -0,0 +1,452 @@
+        // Note: the 'addr' crate also has the 'parse_dns_name' function, which considers
+        // '[1:2:3:4:5:6:7:8]' to be a domain name.
+        match addr::parse_domain_name(s) {
+            Ok(_) => Ok(DomainName(s.to_owned())),
+            Err(_) => Err(DomainNameParseError {
+                erroneous_domain_name: s.to_owned(),
+            }),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("Not a domain name: '{erroneous_domain_name}'")]
+pub struct DomainNameParseError {
+    erroneous_domain_name: String,
+}
+
+impl ResolvableName for DomainName {
+    type ResolvedAddress = IpAddr;
+
+    async fn resolve(&self) -> Result<impl Iterator<Item = IpAddr> + '_, NameResolutionError> {
+        // Note: for historical reasons, std::net::ToSocketAddrs, which is used internally
+        // by tokio::net::lookup_host, expects a pair of (host, port) and resolves it to
+        // SocketAddr, despite the fact that the port number is useless during name resolution.
+        // So, we specify the fake port number 0 here and then ignore it in the result.
+        Ok(tokio::net::lookup_host((self.0.as_str(), 0))
+            .await
+            .map_err(|error| NameResolutionError::CannotResolve {
+                resolvable_name: self.0.clone(),
+                error_str: error.to_string(),
+            })?
+            .map(|socket_addr| socket_addr.ip()))
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum NetworkAddress {
+    Ip(IpAddr),
+    Domain(DomainName),
+}
+
+impl NetworkAddress {
+    pub fn as_ip_address(&self) -> Option<&IpAddr> {
+        match self {
+            Self::Ip(addr) => Some(addr),
+            Self::Domain(_) => None,
+        }
+    }
+}
+
+impl ResolvableName for NetworkAddress {
+    type ResolvedAddress = IpAddr;
+
+    async fn resolve(
+        &self,
+    ) -> Result<impl Iterator<Item = Self::ResolvedAddress> + '_, NameResolutionError> {
+        match self {
+            Self::Ip(ip) => Ok(Either::Left(std::iter::once(*ip))),
+            Self::Domain(domain_name) => Ok(Either::Right(domain_name.resolve().await?)),
+        }
+    }
+}
+
+impl FromStr for NetworkAddress {
+    type Err = NetworkAddressParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(ip) = s.parse::<IpAddr>() {
+            Ok(NetworkAddress::Ip(ip))
+        } else {
+            Ok(NetworkAddress::Domain(s.parse::<DomainName>()?))
+        }
+    }
+}
+
+impl Display for NetworkAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Ip(addr) => {
+                write!(f, "{addr}")
+            }
+            Self::Domain(domain_name) => {
+                write!(f, "{domain_name}")
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, DeserializeFromStr, SerializeDisplay)]
+pub struct NetworkAddressWithOptionalPort {
+    pub address: NetworkAddress,
+    pub port: Option<u16>,
+}
```

Comment:
This is actually something I wanted to do yesterday, but as usual, I never get the time to do these nicities. Thank you.

The problem I wanted to solve is that when specifying a host without a port for binding, it should use the default port. For example, if the default bind address for RPC is 127.0.0.1:3030, and I specify in command line arguments 0.0.0.0, it should automatically become 0.0.0.0:3030.

Reply:
Maybe this should be taken out to be used by all crates?